### PR TITLE
internal/pkg/scorecard: implement plugin system

### PR DESCRIPTION
**Pull request by AlexNPavel**
**Description of the change:** Implement scorecard plugins

**Motivation for the change:** This PR implements support for scorecard plugins as described in this proposal: https://github.com/operator-framework/operator-sdk/blob/master/doc/proposals/scorecard-plugin-system.md

The Basic and OLM test suites will be split out into separate plugins after this PR is merged.

### CODE REVIEW

File: internal/pkg/scorecard/scorecard.go [hunk 1]
```diff
@@ -310,7 +312,63 @@ func runTests() ([]TestSuite, error) {
...
+		// executable files must be in "bin" subdirectory
+		files, err := ioutil.ReadDir("bin")
+		if err != nil {
+			return nil, fmt.Errorf("failed to list files in %s/bin", pluginDir)
+		}
+		for _, file := range files {
+			cmd := exec.Command("bash", "-c", "./bin/"+file.Name())
+			output, err := cmd.CombinedOutput()
+			if err != nil {
+				failedPlugin := scapiv1alpha1.ScorecardOutput{}
```
Comment (estroz) [inline internal/pkg/scorecard/scorecard.go]: This and the below error check code can mostly be refactored into a function.

File: internal/pkg/scorecard/scorecard.go [hunk 2]
```diff
@@ -310,7 +312,63 @@ func runTests() ([]TestSuite, error) {
...
+		if err := os.Chdir(pluginDir); err != nil {
+			return nil, fmt.Errorf("failed to chdir into scorecard plugin directory")
+		}
+		// executable files must be in "bin" subdirectory
+		files, err := ioutil.ReadDir("bin")
+		if err != nil {
+			return nil, fmt.Errorf("failed to list files in %s/bin", pluginDir)
+		}
+		for _, file := range files {
+			cmd := exec.Command("bash", "-c", "./bin/"+file.Name())
```
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: We need a shell to be able to start/execute a shell script. Since we support executing shell scripts, we need to start them through a shell. We could probably use `sh` as well, since the actual language for the shell script should be defined by the first line of the script (like `#!/bin/bash`), but I just assumed we would always run on a system that has bash.

[the remark this answers is not in the document]
Comment (estroz) [reply internal/pkg/scorecard/scorecard.go]: The shebang will take care of shell creation. I think we should not add any system-specific configuration to running binaries, even if `bash` is supported and present in virtually all systems.

[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: Huh. When I was first testing it, it wasn't working without bash. I guess those errors must've been due to something else because it seems to work now. Changed to execute directly.

File: internal/pkg/scorecard/scorecard.go [hunk 3]
```diff
@@ -310,7 +312,63 @@ func runTests() ([]TestSuite, error) {
...
+			output, err := cmd.CombinedOutput()
```
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: Where exactly would we put the log? We provide a per-suite log field for plugins to use so they shouldn't be outputting anything to stdout or stderr except the JSON or potentially fatal errors. I assume we could add it into the main scorecard log (in the final `ScorecardOutput`), but that could get messy pretty quickly

File: internal/pkg/scorecard/scorecard.go [hunk 4]
```diff
@@ -321,42 +363,51 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
+	pluginOutputs, err := runTests()
 	if err != nil {
 		return err
 	}
 	totalScore := 0.0
 	// Update the state for the tests
-	for _, suite := range suites {
-		for idx, res := range suite.TestResults {
-			suite.TestResults[idx] = UpdateState(res)
+	for _, suite := range pluginOutputs {
```
Comment (camilamacedo86) [inline internal/pkg/scorecard/scorecard.go]: A for inside another shows not be the best solution. Note the Big O here. Could it not cause performance issues? Could it not be improved to achieve a better performance?

File: internal/pkg/scorecard/scorecard.go [hunk 5]
```diff
@@ -321,42 +363,51 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
 	}
 	if viper.GetString(OutputFormatOpt) == "human-readable" {
-		for _, suite := range suites {
-			fmt.Printf("%s:\n", suite.GetName())
-			for _, result := range suite.TestResults {
-				fmt.Printf("\t%s: %d/%d\n", result.Test.GetName(), result.EarnedPoints, result.MaximumPoints)
+		numSuites := 0
+		for _, plugin := range pluginOutputs {
+			for _, suite := range plugin.Results {
+				fmt.Printf("%s:\n", suite.Name)
```
Comment (camilamacedo86) [inline internal/pkg/scorecard/scorecard.go]: The same point of attention in the impl. The complexity of this impl.

File: internal/pkg/scorecard/scorecard.go [hunk 6]
```diff
@@ -321,42 +363,51 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
-		for _, suite := range suites {
-			for _, result := range suite.TestResults {
-				for _, suggestion := range result.Suggestions {
-					// 33 is yellow (specifically, the same shade of yellow that logrus uses for warnings)
-					fmt.Printf("\x1b[%dmSUGGESTION:\x1b[0m %s\n", 33, suggestion)
+		for _, plugin := range pluginOutputs {
+			for _, suite := range plugin.Results {
+				for _, result := range suite.Tests {
+					for _, suggestion := range result.Suggestions {
+						// 33 is yellow (specifically, the same shade of yellow that logrus uses for warnings)
```
Comment (camilamacedo86) [inline internal/pkg/scorecard/scorecard.go]: A loop the inside of another loop 4 times. Is not possible to solve this need in a more performative way?

File: internal/pkg/scorecard/scorecard.go [hunk 7]
```diff
@@ -321,42 +363,51 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
 				}
 			}
 		}
 		// Print errors
-		for _, suite := range suites {
-			for _, result := range suite.TestResults {
-				for _, err := range result.Errors {
-					// 31 is red (specifically, the same shade of red that logrus uses for errors)
-					fmt.Printf("\x1b[%dmERROR:\x1b[0m %s\n", 31, err)
+		for _, plugin := range pluginOutputs {
```
Comment (camilamacedo86) [inline internal/pkg/scorecard/scorecard.go]: A loop the inside of another loop 4 times. Is not possible to solve this need in a more performative way?

**Comment by AlexNPavel** [comment]
@camilamacedo86 BigO for these loops is irrelevant to the runtime of the scorecard. Much of the time spent in the scorecard will be on the actual plugins themselves (which includes the Basic and OLM tests), which will likely be waiting on networking. The plugins should be optimized if they are compute heavy. However, the time spent in the loops you mentioned will be on the order of milliseconds (excluding the print syscall time, which we would still have) even for runs with a lot of output and plugins. Those extra couple of milliseconds spent in the loops are irrelevant to the overall runtime of the scorecard and any optimizations there would sacrifice clarity.

**Comment by camilamacedo86** [comment]
@AlexNPavel

IMHO shows that it could be changed/improved. Now it shows irrelevant however, it can grow and the issues started to be faced. The CI tests already take too long to be finished and could not the code be more understandable also and this easier to keep maintained and be improved?

I would try to solve the same need without some many loops inside of other loops. It is just my IMHO and suggestion.

**Comment by shawn-hurley** [comment]
> You make a great work here, it is just a suggestion. Please, feel free to move forward as you decide

I think that if we want to make this better with fewer loops we would need to redesign the structs. 

I think that something we could do in this PR or as a follow up to make this better is:

1. Have the `Plugin` struct responsible for its output. 
2. Allow this function to take a channel of type string. Allow the function to do the loops and create its output and then return it to the channel. 
3. For each plugin calls each of these and then takes the return strings and create the output. 

If that is too much, at the very least lets combine the two different for loops, by creating the output in two string buffers and then combining.


```
for _, suite := range plugin.Results {
				for _, result := range suite.Tests {
					for _, err := range result.Errors {
						// 31 is red (specifically, the same shade of red that logrus uses for errors)
						errorBuffer.Add(fmt.Sprintf("\x1b[%dmERROR:\x1b[0m %s\n", 31, err))
					}
					for _, suggestion := range result.Suggestions {
						// 33 is yellow (specifically, the same shade of yellow that logrus uses for warnings)
						suggestionBuffer.Addfmt.Sprintf("\x1b[%dmSUGGESTION:\x1b[0m %s\n", 33, suggestion))
					}

				}
}
```

/cc @camilamacedo86 @AlexNPavel 

BTW I personally feel this is acceptable to do in a follow-up.

File: internal/pkg/scorecard/scorecard.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/pkg/scorecard/scorecard.go]: Yeah that's a good question. Wdyt of checking if `result.Log` is empty. If so, set `result.Log` to the stderr buffer. If not, log a warning that would be seen in the human-readable output?

File: internal/pkg/scorecard/helpers.go [hunk 8]
```diff
@@ -147,7 +135,7 @@ func TestResultToScorecardTestResult(tr TestResult) scapiv1alpha1.ScorecardTestR
 }
 
 // UpdateState updates the state of a TestResult.
-func UpdateState(res TestResult) TestResult {
+func UpdateState(res scapiv1alpha1.ScorecardTestResult) scapiv1alpha1.ScorecardTestResult {
```
Comment (joelanford) [inline internal/pkg/scorecard/helpers.go]: We have a handful of exported helper functions for the `scapiv1alpha1` types in this package. I'd suggest moving them into the `scapiv1alpha1` package, possibly making them methods on the `scapiv1alpha1` structs (where that makes sense), and also possibly unexporting them if we think they'll only be used on the SDK side of the plugin interface.

This could be in a follow-on PR. Since we're going to reorganize our built-in scorecard tests as plugins, I'd be interested in your thoughts on a design for a plugin library that would make it easy to write a plugin in go.

Comment (AlexNPavel) [reply internal/pkg/scorecard/helpers.go]: Yeah, I'll make a PR to expose some of these. I was initially intending to expose some of this with the JSON PR, but it wasn't quite ready. Now that we have the plugin support in place, we can probably expose the helpers.

File: internal/pkg/scorecard/scorecard.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: That sounds reasonable. I'll do that.

[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: Actually, there's a problem with that. A single plugin can have more than 1 suite. If that's the case, we won't be able to simply add the stderr to a log field.

[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: The way I'll handle this for now is using `log.Warn` for the stderr output, which will print it nicely in the human-readable output or add it to the main log for the json output.

File: internal/pkg/scorecard/helpers.go [hunk 9]
```diff
@@ -161,3 +149,41 @@ func UpdateState(res TestResult) TestResult {
...
+		switch test.State {
+		case scapiv1alpha1.ErrorState:
+			suite.Error++
+		case scapiv1alpha1.PassState:
+			suite.Pass++
+		case scapiv1alpha1.PartialPassState:
+			suite.PartialPass++
+		case scapiv1alpha1.FailState:
+			suite.Fail++
+		}
```
Comment (hasbro17) [inline internal/pkg/scorecard/helpers.go]: We should probably have a default case to catch and log any unknown state. 
Even though it doesn't seem like we'll end up in an unknown state, it's still good practice in case we change it in the future.

File: internal/pkg/scorecard/scorecard.go [hunk 10]
```diff
@@ -310,7 +314,59 @@ func runTests() ([]TestSuite, error) {
...
-	return suites, nil
+	for _, suite := range suites {
+		// convert to ScorecardOutput format
+		// will add log when basic and olm tests are separated into plugins
+		pluginResults = append(pluginResults, TestSuitesToScorecardOutput([]TestSuite{suite}, ""))
+	}
+	// Run plugins
+	pluginDir := viper.GetString(PluginDirOpt)
+	if dir, err := os.Stat(pluginDir); err != nil || !dir.IsDir() {
+		log.Warnf("Plugin directory not found; skipping plugin tests")
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]: Return the result here so we can drop the `else` block. It helps with readability if we fall through.
```Go
if dir, err := os.Stat(pluginDir); err != nil || !dir.IsDir() {
	log.Warnf("Plugin directory not found; skipping plugin tests")
	return pluginResults, nil
}

if err := os.Chdir(pluginDir); err != nil {
...
```

Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]: Also any reason why we're not logging the full error? It'll help with users debug an incorrect path if the error is more explicit:
```Go
	log.Warnf("Plugin directory not found; skipping plugin tests: %v", err)
```

File: internal/pkg/scorecard/scorecard.go [hunk 11]
```diff
@@ -310,7 +314,59 @@ func runTests() ([]TestSuite, error) {
...
+	} else {
+		if err := os.Chdir(pluginDir); err != nil {
+			return nil, fmt.Errorf("failed to chdir into scorecard plugin directory")
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]:
```suggestion
			return nil, fmt.Errorf("failed to chdir into scorecard plugin directory: %v", err)
```

File: internal/pkg/scorecard/scorecard.go [hunk 12]
```diff
@@ -310,7 +314,59 @@ func runTests() ([]TestSuite, error) {
...
+		}
+		// executable files must be in "bin" subdirectory
+		files, err := ioutil.ReadDir("bin")
+		if err != nil {
+			return nil, fmt.Errorf("failed to list files in %s/bin", pluginDir)
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]:
```suggestion
			return nil, fmt.Errorf("failed to list files in %s/bin: %v", pluginDir, err)
```

File: internal/pkg/scorecard/scorecard.go [hunk 13]
```diff
@@ -310,7 +314,59 @@ func runTests() ([]TestSuite, error) {
...
+			if err != nil {
+				pluginResults = append(pluginResults, failedPlugin(
+					fmt.Sprintf("Failed Plugin: %s", file.Name()),
+					fmt.Sprintf("Plugin with file name `%s` failed", file.Name()),
+					fmt.Sprintf("%s:\nStdout: %s\nStderr: %s", err, string(stdout.Bytes()), string(stderr.Bytes()))))
+				// output error to main logger as well for human-readable output
+				log.Errorf("Plugin `%s` failed with error (%v)", file.Name(), err)
+				continue
+			}
+			//parse output and add to suites
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]:
```suggestion
			// parse output and add to suites
```

File: internal/pkg/scorecard/scorecard.go [hunk 14]
```diff
@@ -310,7 +314,59 @@ func runTests() ([]TestSuite, error) {
...
+			stdout := &bytes.Buffer{}
+			cmd.Stdout = stdout
+			stderr := &bytes.Buffer{}
+			cmd.Stderr = stderr
+			err := cmd.Run()
+			if err != nil {
+				pluginResults = append(pluginResults, failedPlugin(
+					fmt.Sprintf("Failed Plugin: %s", file.Name()),
+					fmt.Sprintf("Plugin with file name `%s` failed", file.Name()),
+					fmt.Sprintf("%s:\nStdout: %s\nStderr: %s", err, string(stdout.Bytes()), string(stderr.Bytes()))))
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]: Nit: A little difficult to parse the args being passed to `failedPlugin()`. WDYT about naming the arguments before passing them in:
```Go
name := fmt.Sprintf("Failed Plugin: %s", file.Name())
description := fmt.Sprintf("Plugin with file name `%s` failed", file.Name())
log := fmt.Sprintf("%s:\nStdout: %s\nStderr: %s", err, string(stdout.Bytes()), string(stderr.Bytes()))
pluginResults = append(pluginResults, failedPlugin(name, description, log))
```

Comment (hasbro17) [reply internal/pkg/scorecard/scorecard.go]: Same for Lines 355-358

File: internal/pkg/scorecard/scorecard.go [hunk 15]
```diff
@@ -321,42 +377,51 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
 	// Update the state for the tests
-	for _, suite := range suites {
-		for idx, res := range suite.TestResults {
-			suite.TestResults[idx] = UpdateState(res)
+	for _, suite := range pluginOutputs {
+		for idx, res := range suite.Results {
+			suite.Results[idx] = UpdateSuiteStates(res)
 		}
 	}
 	if viper.GetString(OutputFormatOpt) == "human-readable" {
```
Comment (hasbro17) [inline internal/pkg/scorecard/scorecard.go]: While it's not in this PR but can we make the output format flag values(`human-readable` and `json` ) into constants?
Right now they're appearing in several places as string literals.

Should be a small change, or you can make a separate PR for it.

**Comment by hasbro17** [comment]
Looks good overall. Just some small nits.

File: internal/pkg/scorecard/helpers.go (hunk 9 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/helpers.go]: I'll do this in a future PR. The issue is that there is no way to really handle this cleanly right now. One way to fix this would be to add another field to the SuiteResults of `invalid`. That way we can indicate to users that a plugin returned an invalid state. Another option (which is what I want to do in a future PR) is to change the state from whatever incorrect state it was in to `error` and add a string to the errors array indicating that the scorecard marked this test as failed due to an invalid state.


Technically though, due to the way the helpers are set up (we calculate the state using the earned vs max points unless the state is error), the only situation where this could occur (currently) would be if the earned points are higher than the max. I'd be fine marking that as an error.

**Comment by AlexNPavel** [comment]
@hasbro17 I am going to make a PR with all the doc changes today
